File: algorithms/zero_matrix.rb
# CtCI Edition 6 Problem 1.8
# Zero Matrix: Write an algorithm such that if an element in an MxN matrix is 0, its entire row and
# column are set to O.

# Matrix is NxM
class Matrix
  def initialize(matrix_array)
    @matrix = matrix_array
    @rows = @matrix.length - 1
    @cols = @matrix[0].length - 1
  end

  # returns a new zeroed Matrix
  def zero_matrix
    @zeroed_rows = Set.new
    @zeroed_cols = Set.new
    matrix = @matrix.dup

    find_zeros
    zeroize_rows(matrix)
    zeroize_cols(matrix)

    Matrix.new matrix
  end

  def ==(other)
    self.class == other.class && to_a == other.to_a
  end

  def to_a
    @matrix
  end

  private

  # put zero positions into @zeroed_rows and @zeroed_cols
  def find_zeros
    0.upto(@rows).each do |row|
      0.upto(@cols).each do |col|
        if (@matrix[row][col]).zero?
          @zeroed_rows << row
          @zeroed_cols << col
        end
      end
    end
  end

  # put a zero at every position in a rows in @zeroed_rows
  def zeroize_rows(matrix)
    @zeroed_rows.each do |row|
      0.upto(@cols).each do |col|
        matrix[row][col] = 0
      end
    end
  end

  # put a zero at every position in a column in @zeroed_cols
  def zeroize_cols(matrix)
    @zeroed_cols.each do |col|
      0.upto(@rows).each do |row|
        matrix[row][col] = 0
      end
    end
  end
end

File: algorithms/zero_matrix.py
#!/usr/bin/env python3
"""Python solution for CtCI 6th Edition Problem 1.8

Write an algorithm such that if an element in an MxN matrix is 0, its entire
row and column are set to O.

No element is 0
>>> zero_matrix([['q', 'w', 'e'], ['a', 's', 'd'], ['z', 'x', 'c']])
[['q', 'w', 'e'], ['a', 's', 'd'], ['z', 'x', 'c']]

One element is 0
>>> zero_matrix([['q', 'w', 'e'], ['a', '0', 'd'], ['z', 'x', 'c']])
[['q', '0', 'e'], ['0', '0', '0'], ['z', '0', 'c']]

Two elements are 0 in different row/column
>>> zero_matrix([['1', '2', '3', '0'], ['q', 'w', 'e', 'r'], ['a', 's', 'd', 'f'], ['0', 'x', 'c', 'v']])
[['0', '0', '0', '0'], ['0', 'w', 'e', '0'], ['0', 's', 'd', '0'], ['0', '0', '0', '0']]

Two elements are 0 in the same column
>>> zero_matrix([['1', '0', '3', '4'], ['q', 'w', 'e', 'r'], ['a', '0', 'd', 'f'], ['z', 'x', 'c', 'v']])
[['0', '0', '0', '0'], ['q', '0', 'e', 'r'], ['0', '0', '0', '0'], ['z', '0', 'c', 'v']]

Solution looks at each element in each list, more than one, making it O(E*L)
"""

def zero_matrix(matrix):
    zero_coordinates = find_zeros(matrix)
    if len(zero_coordinates['rows']) > 0:
        for list_number, list in enumerate(matrix):
            for element_number, element in enumerate(list):
                if (   list_number    in zero_coordinates['rows']
                    or element_number in zero_coordinates['columns']):
                        matrix[list_number][element_number] = '0'
    return matrix


def find_zeros(matrix):
    """Returns a dict  with list of rows and columns that contain zeros
    within the matrix

    >>> zero_coordinates = find_zeros([['0', 'w', 'e'], ['a', 's', 'd'], ['z', 'x', '0']])
    >>> sorted(zero_coordinates.items())
    [('columns', [0, 2]), ('rows', [0, 2])]
    """
    rows = []
    columns = []
    coordinates = {'rows': rows, 'columns': columns}
    for list_number, list in enumerate(matrix):
        for element_number, element in enumerate(list):
            if element == '0':
                coordinates['rows'].append(list_number)
                coordinates['columns'].append(element_number)
    return coordinates
